agent/lib_aws_sdk_php.c

```diff
@@ -14,10 +14,411 @@
+    command_arg_array = nr_php_zend_hash_index_find(Z_ARRVAL_P(param_array), 0);
+    if (nr_php_is_zval_valid_array(command_arg_array)) {
+      zval* queueurl_arg = nr_php_zend_hash_find(Z_ARRVAL_P(command_arg_array),
+                                                 command_arg_name);
+
+      if (nr_php_is_zval_non_empty_string(queueurl_arg)) {
+        command_arg_value = nr_strdup(Z_STRVAL_P(queueurl_arg));
+      }
+    }
+  }
+
+  nr_php_arg_release(&param_array);
+  return command_arg_value;
+}
+
+/*
+ * For Aws/AwsClient::__call see
+ * https://github.com/aws/aws-sdk-php/blob/master/src/AwsClientInterface.php
+ * ALL
+ * client commands are handled by this function, so it is the start and end of
+ * any command. Creates and executes a command for an operation by name.
+ * When a class command isn't explicitly created as a function, the __call class
+ * handles the invocation.  This means all AWS Client Service commands are
+ * handled by this call. Any invocation starts when this function starts, and
+ * ends when it ends.  This function decodes the command name, determines the
+ * appropriate args, decodes the args, generates a guzzle request to send to the
+ * AWS Service, gets the guzzle response from the AWS Service, and bundles that
+ * response into an AswResult to return.
+ *
+ * @param string $name      Name of the command to execute.
+ * @param array  $arguments Arguments to pass to the getCommand method.
+ *
+ * @return ResultInterface
+ * @throws \Exception
+ */
+
+NR_PHP_WRAPPER(nr_aws_client_call_before) {
+  (void)wraprec;
+  nr_segment_t* segment = NULL;
+  /*
+   * Start a segment in case it needs to become an external, message, or
+   * datastore segment.
+   */
+  segment = nr_segment_start(NRPRG(txn), NULL, NULL);
+  if (NULL != segment) {
+    segment->wraprec = auto_segment->wraprec;
+  }
+}
+NR_PHP_WRAPPER_END
+
+NR_PHP_WRAPPER(nr_aws_client_call) {
+  (void)wraprec;
+
+  const char* klass = NULL;
+  char* command_name_string = NULL;
+  char* real_class_and_command = NULL;
+  nr_segment_t* segment = NULL;
+  zend_class_entry* class_entry = NULL;
+
+  class_entry = Z_OBJCE_P(nr_php_execute_scope(execute_data));
+  klass = nr_php_class_entry_name(class_entry);
+  command_name_string
+      = nr_lib_aws_sdk_php_get_command_name(NR_EXECUTE_ORIG_ARGS);
+
+  if (NULL != command_name_string && NULL != klass) {
+    if (nr_striendswith(klass, nr_php_class_entry_name_length(class_entry),
+                        AWS_SDK_PHP_SQSCLIENT_CLASS,
+                        AWS_SDK_PHP_SQSCLIENT_CLASS_LEN)) {
+      nr_lib_aws_sdk_php_sqs_handle(auto_segment, command_name_string,
+                                    NR_EXECUTE_ORIG_ARGS);
+    }
+  }
+
+  if (NR_SEGMENT_CUSTOM == auto_segment->type) {
+    /*
+     * It wasn't set to any of the segments that would autoclose the segment so
+     * we need to end the segment that we started in the before wrapper to
+     * handle external,datastore,message segments
```

Comment:
```suggestion
```suggestion
   * We need to end the segment that we started in the 'before' wrapper if it
   * wasn't handled and ended by the handling function. Handling function would
   * have changed the segment type from from default (`NR_SEGMENT_CUSTOM`) if it
   * ended it.
```